src/WebJobs.Script/Metrics/HostMetricsProvider.cs

```diff
@@ -93,33 +104,29 @@ private void AddOrUpdateMetricsCache(string key, long value)
                 return null;
             }
 
-            try
+            var functionActivityStatusProvider = _serviceProvider.GetScriptHostServiceOrNull<IFunctionActivityStatusProvider>();
+            if (functionActivityStatusProvider is not null)
             {
-                var functionActivityStatusProvider = _serviceProvider.GetScriptHostServiceOrNull<IFunctionActivityStatusProvider>();
-                if (functionActivityStatusProvider is not null)
-                {
-                    var functionActivityStatus = functionActivityStatusProvider.GetStatus();
-                    AddOrUpdateMetricsCache(HostMetrics.ActiveInvocationCount, functionActivityStatus.OutstandingInvocations);
-                }
-
-                return new Dictionary<string, long>(_metricsCache);
+                var functionActivityStatus = functionActivityStatusProvider.GetStatus();
+                AddOrUpdateMetricsCache(HostMetrics.ActiveInvocationCount, functionActivityStatus.OutstandingInvocations);
             }
-            finally
+
+            IReadOnlyDictionary<string, long> metrics;
+
+            lock (_lock)
             {
-                PurgeMetricsCache();
+                metrics = new Dictionary<string, long>(_metricsCache);
+                _metricsCache.Clear();
```

Comment:
@fabiocav I did look into Interlocked.Exchange but it would have meant having to return a concurrent dictionary instead of an `IReadOnlyDictionary` and I would like to avoid the extra allocation of converting the empty concurrent dict to a read only one. Therefore I went for this approach instead. Thoughts?